### PR TITLE
Forward Mode CallInst

### PR DESCRIPTION
**Pull request by tgymnich**
**TODO**:
- [x] handle CallInst
- [x] Checks and debug messages
- [x] Tests

### CODE REVIEW

File: enzyme/Enzyme/AdjointGenerator.h [hunk 1]
```diff
@@ -1346,6 +1346,8 @@ class AdjointGenerator
   }
 
   Value *lookup(Value *val, IRBuilder<> &Builder) {
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:1348]: I'd move this assertion within GradientUtils

Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h]: done c864b891711b8aa4953e97f347ca2176364251ef

Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h]: I might put this in a separate PR, since there is a lot of breakage with existing forward mode...

Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h]: #306

File: enzyme/Enzyme/AdjointGenerator.h [hunk 2]
```diff
@@ -6654,10 +6654,115 @@ class AdjointGenerator
...
+      diffes->setDebugLoc(gutils->getNewFromOriginal(orig->getDebugLoc()));
+#if LLVM_VERSION_MAJOR >= 9
+      for (auto pair : gradByVal) {
+        diffes->addParamAttr(
+            pair.first,
+            Attribute::getWithByValType(diffes->getContext(), pair.second));
+      }
+#endif
+
+      if (!diffes->getType()->isVoidTy()) {
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6751]: Should this not be a check whether the value returned by the call is constant?

Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:6751]: yes!

**Review by wsmoses — APPROVED** [review]

**Pull request merged**